File: leetcode/python/653. 两数之和 IV - 输入 BST.py
#!/bin/bash/python
"""
653. 两数之和 IV - 输入 BST

思路一：
中序遍历生成hashSet

思路二：
中序遍历生成有序数组，双指针

思路三：
直接递归（待补充）
"""

class TreeNode(object):
	def __init__(self, val=0, left=None, right=None):
		self.val = val
		self.left = left
		self.right = right


class Solution(object):
	def findTarget(self, root, k):
		"""
		hashSet
		"""
		d = dict()
		def mid(root):
			if root == None:
				return
			d[root.val] = True
			mid(root.left)
			mid(root.right)
		mid(root)
		for key in d:
			if key * 2 != k and (k - key) in d:
				return True
		return False
	
	def findTarget2(self, root):
		"""
		中序遍历生成有序数组+双指针
		"""
		node = []
		def mid(root):
			if root == None:
				return
			mid(root.left)
			node.append(root.val)
			mid(root.right)
		mid(root)
		l, r = 0, len(node) - 1
		while l < r:
			ssum = node[l] + node[r]
			if ssum == k:
				return True
			elif ssum < k:
				l += 1
			else:
				r -= 1
		return False

File: leetcode/cpp/653. 两数之和 IV - 输入 BST.cpp
#include <vector>
#include <map>

typedef struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int val=0) : val(val), left(nullptr), right(nullptr) {}
} TreeNode;

class Solution {
public:
    // 中序遍历生成哈希
	void mid(TreeNode* root, std::map<int, bool>& map) {
		if (root == nullptr) {
			return;
		}
		map[root->val] = true;
		mid(root->left, map);
		mid(root->right, map);
		return;
	}

	// 中序遍历生成有序数组
	void mid(TreeNode* root, std::vector<int>& vec) {
		if (root == nullptr) {
			return;
		}
		mid(root->left, vec);
		vec.push_back(root->val);
		mid(root->right, vec);
		return;
	}

	// 哈希法
	bool findTarget(TreeNode* root, int k) {
		std::map<int, bool> map;
		mid(root, map);
		for (auto it = map.begin(); it != map.end(); it++) {
			int val = it->first;
			if (val * 2 != k && map.count(k - val)) {
				return true;
			}
		}
		return false;
	}

	// 双指针
	bool findTarget2(TreeNode* root, int k) {
		std::vector<int> vec;
		mid(root, vec);
		int l = 0, r = vec.size() - 1;
		while (l < r) {
			int sum = vec[l] + vec[r];
			if (sum == k) {
				return true;
			} else if (sum < k) {
				l += 1;
			} else {
				r -= 1;
			}
		}
		return false;
	}
};